File: vpbuf/examples/uno/rel1/cpp/src/vp_uno.cc
#pragma once
namespace vp_uno {

	bool version_check(long ver)
	{
		return (ver < 1 || ver > 1) ? false : true;
	}
	long get_high_version() {
		return 1;
	}
	long get_low_version() {
		return 1;
	}

	void write_Header(write_context &ctx, Header &payload)
	{
		write_int(ctx, payload.version);
		write_string(ctx, payload.tag);
	}

	void write_Value(write_context &ctx, Value &);
	void write_Reverse(write_context &ctx, Reverse &);
	void write_Skip(write_context &ctx, Skip &);
	void write_DrawTwo(write_context &ctx, DrawTwo &);

	void write_Card(write_context &ctx, Card &payload)
	{
		const std::type_info& tid = typeid(payload);
		if (std::type_index(tid) == std::type_index(typeid(Value))) {
			write_int(ctx,0);
			write_Value(ctx, dynamic_cast<Value &> (payload));
		}
		if (std::type_index(tid) == std::type_index(typeid(Reverse))) {
			write_int(ctx,1);
			write_int(ctx,0);
			write_Reverse(ctx, dynamic_cast<Reverse &> (payload));
		}
		if (std::type_index(tid) == std::type_index(typeid(Skip))) {
			write_int(ctx,1);
			write_int(ctx,1);
			write_Skip(ctx, dynamic_cast<Skip &> (payload));
		}
		if (std::type_index(tid) == std::type_index(typeid(DrawTwo))) {
			write_int(ctx,1);
			write_int(ctx,2);
			write_DrawTwo(ctx, dynamic_cast<DrawTwo &> (payload));
		}
		write_int(ctx, payload.id);
	}

	void write_Value(write_context &ctx, Value &payload)
	{
		write_int(ctx, payload.value);
		write_string(ctx, payload.color);
	}

	void write_Action(write_context &ctx, Action &payload)
	{
		write_string(ctx, payload.color);
	}

	void write_Reverse(write_context &ctx, Reverse &payload)
	{
		write_Action(ctx, payload);
	}

	void write_Skip(write_context &ctx, Skip &payload)
	{
		write_Action(ctx, payload);
	}

	void write_DrawTwo(write_context &ctx, DrawTwo &payload)
	{
		write_Action(ctx, payload);
	}

	void write_Deck(write_context &ctx, Deck &payload)
	{
		write_int(ctx, payload.cards.size());
		for (auto ii = payload.cards.begin();ii != payload.cards.end(); ii++)
			write_Card(ctx, *(*ii));
	}

	void read_Header(read_context &ctx, Header &payload)
	{
		read_int(ctx, payload.version);
		read_string(ctx, payload.tag);
	}

	Value* read_Value(read_context &ctx);
	Action* read_Action(read_context &ctx);

	Card* read_Card(read_context &ctx)
	{
		Card* payload_ptr; int t; read_int(ctx, t);
		if (t == 0)
			payload_ptr = read_Value(ctx);
		if (t == 1)
			payload_ptr = read_Action(ctx);
		auto &payload = *payload_ptr;
		read_int(ctx, payload.id);
		return payload_ptr;
	}

	Value *read_Value(read_context &ctx)
	{
		auto *payload_ptr = new Value;
		auto &payload = *payload_ptr;
		read_int(ctx, payload.value);
		read_string(ctx, payload.color);
		return payload_ptr;
	}

	void read_Value(read_context &ctx, Value &payload)
	{
		read_int(ctx, payload.value);
		read_string(ctx, payload.color);
	}

	Reverse* read_Reverse(read_context &ctx);
	Skip* read_Skip(read_context &ctx);
	DrawTwo* read_DrawTwo(read_context &ctx);

	Action* read_Action(read_context &ctx)
	{
		Action* payload_ptr; int t; read_int(ctx, t);
		if (t == 0)
			payload_ptr = read_Reverse(ctx);
		if (t == 1)
			payload_ptr = read_Skip(ctx);
		if (t == 2)
			payload_ptr = read_DrawTwo(ctx);
		auto &payload = *payload_ptr;
		read_string(ctx, payload.color);
		return payload_ptr;
	}

	Reverse *read_Reverse(read_context &ctx)
	{
		auto *payload_ptr = new Reverse;
		auto &payload = *payload_ptr;
		return payload_ptr;
	}

	void read_Reverse(read_context &ctx, Reverse &payload)
	{
	}

	Skip *read_Skip(read_context &ctx)
	{
		auto *payload_ptr = new Skip;
		auto &payload = *payload_ptr;
		return payload_ptr;
	}

	void read_Skip(read_context &ctx, Skip &payload)
	{
	}

	DrawTwo *read_DrawTwo(read_context &ctx)
	{
		auto *payload_ptr = new DrawTwo;
		auto &payload = *payload_ptr;
		return payload_ptr;
	}

	void read_DrawTwo(read_context &ctx, DrawTwo &payload)
	{
	}

	void read_Deck(read_context &ctx, Deck &payload)
	{
		int count_cards;
		read_int(ctx, count_cards);
		for (auto i = 0; i < count_cards; i++) {
			auto *pod = read_Card(ctx);
			payload.cards.push_back(pod);
		}
	}

	void init_reorder_map(map<string, ReorderCog *> &rmap, int ver) {
	}

}

File: vpbuf/examples/uno/rel1/python/vp_uno/vp_uno.py
from vp_uno.persist import *

def version_check(ver):
	if (ver < 1 or ver > 1):
		return False
	else:
		return True

def get_high_version():
	return 1

def get_low_version():
	return 1

def init_reorder_map(map, ver, seed):
	pass

def write_str(ctx, payload):
    ctx.encoder.writeString(payload) # amf3

def write_int(ctx, payload):
    ctx.encoder.writeInteger(payload) # amf3

def write_Header(ctx, payload):
	write_int(ctx, payload.version)
	write_str(ctx, payload.tag)

def write_Card(ctx, payload):
	c = type(payload).__name__
	if (c == "Value"):
		write_int(ctx,0)
		write_Value(ctx, payload)
	if (c == "Reverse"):
		write_int(ctx,1)
		write_int(ctx,0)
		write_Reverse(ctx, payload)
	if (c == "Skip"):
		write_int(ctx,1)
		write_int(ctx,1)
		write_Skip(ctx, payload)
	if (c == "DrawTwo"):
		write_int(ctx,1)
		write_int(ctx,2)
		write_DrawTwo(ctx, payload)
	write_int(ctx, payload.id)


def write_Value(ctx, payload):
	write_int(ctx, payload.value)
	write_str(ctx, payload.color)

def write_Action(ctx, payload):
	write_str(ctx, payload.color)


def write_Reverse(ctx, payload):
	write_Action(ctx, payload)

def write_Skip(ctx, payload):
	write_Action(ctx, payload)

def write_DrawTwo(ctx, payload):
	write_Action(ctx, payload)

def write_Deck(ctx, payload):
	count = len(payload.cards)
	write_int(ctx, count)
	i = 0
	while (i < count):
		write_Card(ctx, payload.cards[i])
		i = i + 1

def read_str(ctx):
   t = ctx.decoder.readInteger() # amf3
   assert(t == 6)
   return ctx.decoder.readString()

def read_int(ctx):
   t = ctx.decoder.readInteger() # amf3
   assert(t == 4)
   return ctx.decoder.readInteger()

def read_Header(ctx):
	payload = Header()
	payload.version = read_int(ctx)
	payload.tag = read_str(ctx)
	return payload

def read_Card(ctx):
	t = read_int(ctx)
	if (t == 0):
		payload = read_Value(ctx)
	if (t == 1):
		payload = read_Action(ctx)
	payload.id = read_int(ctx)
	return payload

def read_Value(ctx):
	payload = Value()
	payload.value = read_int(ctx)
	payload.color = read_str(ctx)
	return payload

def read_Action(ctx):
	t = read_int(ctx)
	if (t == 0):
		payload = read_Reverse(ctx)
	if (t == 1):
		payload = read_Skip(ctx)
	if (t == 2):
		payload = read_DrawTwo(ctx)
	payload.color = read_str(ctx)
	return payload

def read_Reverse(ctx):
	payload = Reverse()
	return payload

def read_Skip(ctx):
	payload = Skip()
	return payload

def read_DrawTwo(ctx):
	payload = DrawTwo()
	return payload

def read_Deck(ctx):
	payload = Deck()
	payload.cards = []
	count = read_int(ctx)
	i = 0
	while (i < count):
		t = read_Card(ctx)
		payload.cards.append(t)
		i = i + 1
	return payload

